### ISSUE TITLE
Suggest: Add plugin function

### ISSUE THREAD

**Issue by cqjjjzr** [MultiMC/MultiMC5#1306]
Hi there.
I'm now using MultiMC and I think it is a very awesome launcher. But when I play on some server, the mod will update and change frequently. And there is some function addable. So I suggest you sincerely add plugin function. You can write some APIs to let developer to write plugin on it. And you can make plugin is loadable to MultiMC. Then, i think the function of MMC can get a lot of improvement.

CharlieJiang
P.S. If i wrote something about this, may I commit it via PullRequset?

**Comment by 02JanDal** [comment]
The short answer is: no.

The longer answer is: we don't want to provide a plugin API for third parties, because that would make us less agile. The codebase is still going through large refactors regularly, and if we have an API we need to stay somewhat compatible with it, which would make refactors a lot harder, if not impossible. We are slowly moving towards restructuring the internal code to be "plugin" based though, but in a static sense (internal components isolated like plugins, and possibly also loaded somewhat like a plugin would, but still everything in the same build tree).

If you give us some more information on the feature you'd like to add we could consider that, and if it's something we'd like to include a PR would be very much appreciated :D

**Comment by cqjjjzr** [comment]
@02JanDal arr yes.
First, what is the "refactor" you mean? the function make programming more convenient? Yes it maybe will make something complex if add plugin.
And "plugin" I mean is a external function for MMC that wrote by 3rd party and add by user. in others, I mean this is like mods and plugins for Bukkit. Maybe it will cause error, performance, but it is in except or canbe avoid or endure.
Saying "agility", sure it will comes with lower agile. But if user really need it, as mod-auto-update, or UI custom, the less agile is canbe endure.

Another way, you can make MultiMC like Eclipse or what, anything on plugin.but it will almost rewrite whole project, so this just a idea. LOL
Sincerely.

**Comment by 02JanDal** [comment]
"Refactor" = restructuring code

Forge mods and bukkit/sponge are intended from the ground up to work with plugins, and wouldn't make sense without plugins. They are not comparable to a project like MultiMC.

Try to explain what you want added, not how ("I want X", not "I want plugins to implement X"). It sounds like you want functionality to update mods easily, is that correct?

**Comment by peterix** [comment]
I see no reason to allow plugins or to maintain API/ABI compatibility.

**Comment by peterix** [comment]
Let me rephrase that.

There is no API. There will be no API in quite a while. Rather than acting as a 3rd party, you should build plugins (cleanly separated pieces of functionality) within the existing codebase. When there is enough of those, we'll have some idea of what sort of API is necessary. We are slowly moving the code that way, but it's really not a priority. If you can push the code that way, it will happen sooner. Componentize the existing features, maybe?

Let's call it components :)

**Comment by 02JanDal** [comment]
This is something we're moving towards anyway (even though it probably won't be exactly as @cqjjjzr described).
